NLP/main/extract_display_quotes.py

```diff
@@ -0,0 +1,596 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+'''
+Created on Mon Jul  4 14:55:46 2022
+
+@author: sjuf9909
+
+This code has been adapted (with permission) from the GenderGapTracker GitHub page 
+(https://github.com/sfu-discourse-lab/GenderGapTracker/tree/master/NLP/main)
+and modified to run on a Jupyter Notebook.
+
+The quotation tool’s accuracy rate is evaluated in the below article:
+The Gender Gap Tracker: Using Natural Language Processing to measure gender bias in media
+(https://doi.org/10.1371/journal.pone.0245533)
+'''
+
+# import required packages
+import os
+import io
+import sys
+import codecs
+import logging
+import traceback
+from collections import Counter
+
+# matplotlib: visualization tool
+from matplotlib import pyplot as plt
+
+# pandas: tools for data processing
+import pandas as pd
+
+# spaCy and NLTK: natural language processing tools for working with language/text data
+import spacy
+from spacy import displacy
+from spacy.tokens import Span
+import nltk
+nltk.download('punkt')
+from nltk import Tree
+from nltk.tokenize import sent_tokenize
+
+# ipywidgets: tools for interactive browser controls in Jupyter notebooks
+import ipywidgets as widgets
+from ipywidgets import Layout
+from IPython.display import display, Markdown, clear_output
+
+# import the quote extractor tool
+# Source: https://github.com/sfu-discourse-lab/GenderGapTracker
+from quote_extractor import extract_quotes
+from config import config
+import utils
+
+
+class QuotationTool():
+    '''
+    Interactive tool for extracting and displaying quotes in a text
+    '''
+    
+    def __init__(self):
+        '''
+        Initiate the QuotationTool
+        '''
+        # initiate the app_logger
+        self.app_logger = utils.create_logger('quote_extractor', log_dir='logs', 
+                                         logger_level=logging.INFO,
+                                         file_log_level=logging.INFO)
+        
+        # download spaCy's en_core_web_lg, the pre-trained English language tool from spaCy
+        print('Loading spaCy language model...')
+        self.nlp = spacy.load('en_core_web_lg')
+        print('Finished loading.')
+        
+        # initiate variables to hold texts and quotes in pandas dataframes
+        self.text_df = None
+        self.quotes_df = None
+        
+        # initiate other variables
+        self.html = None
+        self.current_text = None
+        
+    
+    def upload_files(self):
+        '''
+        Upload text or excel files as input to the QuotationTool
+        '''
+        # widget to upload .txt files
+        uploader_text = widgets.FileUpload(
```

Comment:
I think we could let user upload either .txt or .xlsx file, and call the right data loader based on the file extension.

Currently we are not allowing multiple excel uploading, but if they are with the same header, I don't see why we can't concat them as a dataframe then process.

Reply:
The file upload section has been updated with a tab widget to select text or excel. for excel, I think it makes sense for the user to add more data into their excel spreadsheet before uploading, we just need to educate them on how to use this tool. otherwise, we'll keep updating the tool for too many diff scenarios